collector/sysctl_bsd.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build freebsd dragonfly
+// +build !nomeminfo
+
+package collector
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/sys/unix"
+)
+
+type bsdSysctlType uint8
+
+// BSD-specific sysctl value types.  There is an impedience mismatch between
+// native C types, e.g. int vs long, and the golang unix.Sysctl variables
+
+const (
+	// Default to uint32.
+	bsdSysctlTypeUint32 bsdSysctlType = iota
+	bsdSysctlTypeUint64
+)
+
+// Contains all the info needed to map a single bsd-sysctl to a prometheus value.
+type bsdSysctl struct {
+	// Prometheus name
+	name string
+
+	// Simple prometheus description
+	description string
+
+	// Cached prometheus description
+	cached_descriptor *prometheus.Desc
```

Comment:
I feel like this is a good target for the constructor, i.e. initializing values, and is how stat_linux.go is doing it (but included in the stat struct).   I feel like there's a potential for many sysctls to be exported, and checked on a short interval.  It feels like needless thrashing about, but you're right, it may not exist at all.  I guess that there is similar precedent in the code base?  It would be strange to be inconsistent about the correct way to initialize the descriptions.  On every read, or on initialization.  Happy to go either way on that.